### PR TITLE
Add support for Camera Control

### PR DESCRIPTION
**Pull request by lbegani**
This change is to enable users to controls the camera parameters.

- Details on how the camera parameters will be exposed by the camera(drone) to ground control station, how GCS will communicate with drone to set/get camera params can be found here - http://discuss.px4.io/t/mavlink-camera-api-discussion/3278/31.

- Changes on the QGC for this feature can be found here - https://github.com/mavlink/qgroundcontrol/tree/cameraControl

### CODE REVIEW

File: src/mavlink_server.cpp [hunk 1]
```diff
@@ -303,3 +529,48 @@ void MavlinkServer::stop()
...
+            break;
+        }
+    }
+
+    return;
+}
+
+CameraComponent *MavlinkServer::getCameraComponent(int compID)
+{
+    if (compID < MAV_COMP_ID_CAMERA || compID > MAV_COMP_ID_CAMERA6)
```
Comment (otaviobp) [inline src/mavlink_server.cpp]: Are you using a different componentID for each camera? What if we have more than 6 cameras?

File: src/CameraComponent_V4L2.cpp [hunk 2]
```diff
@@ -0,0 +1,467 @@
...
+
+CameraComponent_V4L2::CameraComponent_V4L2(std::string devicepath, std::string uri)
+    : CameraComponent()
+{
+
+    log_debug("%s path:%s uri:%s", __func__, devicepath.c_str(), uri.c_str());
+    // TODO:: get the supported paramters from the camera device
+    // Initialize the supported and default values
+    dev_path = devicepath;
+    strcpy((char *)camInfo.cam_definition_uri, uri.c_str());
```
Comment (otaviobp) [inline src/CameraComponent_V4L2.cpp]: strcpy are not safe. you can have a buffer overflow here if uri.size() > sizeof(camInfo.cam_definition_uri).

File: src/CameraComponent_V4L2.cpp [hunk 3]
```diff
@@ -0,0 +1,467 @@
...
+    std::string value = camParam.getParameter(param_id);
+    strcpy(param_value, value.c_str());
+    return 1;
+}
+
+int CameraComponent_V4L2::setParam(const char *param_id, const char *param_value, int param_type)
+{
+    int ret = 1;
+    mavlink_param_union_t u;
+    memcpy(&u.param_float, param_value, sizeof(float));
```
Comment (otaviobp) [inline src/CameraComponent_V4L2.cpp]: This looks like a possible overflow or invalid bytes copied when size of the param_value is < than sizeof(float). Is this going to work for uint8?

File: src/CameraComponent_V4L2.cpp [hunk 4]
```diff
@@ -0,0 +1,467 @@
...
+    return 0;
+}
+
+int CameraComponent_V4L2::saveParameter(const char *param_id, float param_value)
+{
+    char str[128];
+    mavlink_param_union_t u;
+    u.param_float = param_value;
+    memcpy(&str[0], &u.param_float, sizeof(float));
+    camParam.setParameter(param_id, str);
```
Comment (otaviobp) [inline src/CameraComponent_V4L2.cpp]: Unless param_value is a string disguised as a float, this is not going to work.
memcpy is not going to convert from a float to a char[] and then you call setParameter that is expecting a std::string. If you are going to store the parameter as a float it is better to use a float vector and not a std::string vector. If you are going to use it as a string, it doesn't make sense to cast it to float.
And anyway, if you are going to use strings, you need to ensure '\0' is appended at the end of the string.

File: src/CameraParameters.h [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+        ID_PIXEL_FORMAT_YUV422I,
+        ID_PIXEL_FORMAT_YUV420P,
+        ID_PIXEL_FORMAT_RGB565,
+        ID_PIXEL_FORMAT_RGBA8888
+    };
+
+    // TODO :: Make exhaustive list of parameters and its possible values
+private:
+    void initParamIdType();
+    void set(std::map<std::string, std::string> &pMap, std::string key, std::string value);
```
Comment (otaviobp) [inline src/CameraParameters.h]: You don't need to create a function to call on method in one object. Just call the method in that object. In this case you can use:

pMap[key] = value;

and you won't need the set method

File: src/CameraParameters.h [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+    void set(std::map<std::string, std::pair<int, int>> &pMap, std::string key,
+             std::pair<int, int> value);
+    std::string get(std::map<std::string, std::string> &pMap, std::string key);
```
Comment (otaviobp) [inline src/CameraParameters.h]: same for get

File: src/CameraParameters.h (hunk 5 above)
Comment (lbegani) [inline src/CameraParameters.h]: Done. I initially created function to be a placeholder for parameter validity checks etc but seems that would be redundant.

File: src/CameraParameters.h (hunk 6 above)
Comment (lbegani) [inline src/CameraParameters.h]: Done

File: src/CameraComponent_V4L2.cpp (hunk 4 above)
Comment (lbegani) [inline src/CameraComponent_V4L2.cpp]: I have added change to explicitly convert byte array to string. I tested it by setting a value to a parameter and getting the value again from the map to see if they match. It works for float, int32, uint32 and uint8.

File: src/CameraComponent_V4L2.cpp (hunk 3 above)
Comment (lbegani) [inline src/CameraComponent_V4L2.cpp]: Fixed it, now just copy sizeof(uint8).

File: src/CameraComponent_V4L2.cpp (hunk 2 above)
Comment (lbegani) [inline src/CameraComponent_V4L2.cpp]: Fixed by adding checks for  sizes

File: src/mavlink_server.cpp (hunk 1 above)
Comment (lbegani) [inline src/mavlink_server.cpp]: Currently 7 component IDs are allotted for camera components in MAVLink. Currently its not an issue as we have less than 7 cameras. Once we hit the number we may need to allot more IDs in the protocol. Technically, there should not be an issue using ID beyond MAV_COMP_ID_CAMERA6 but if QGC put checks for not receiving heartbeats from such out of range components then change in protocol is the only way.

File: src/mavlink_server.cpp [hunk 7]
```diff
@@ -213,32 +285,183 @@ void MavlinkServer::_handle_camera_set_video_stream_settings(const struct sockad
...
+    //          msg->msgid);
 
     if (msg->msgid == MAVLINK_MSG_ID_COMMAND_LONG) {
         mavlink_command_long_t cmd;
         mavlink_msg_command_long_decode(msg, &cmd);
+        log_debug("Command received: (sysid: %d compid: %d msgid: %d)", cmd.target_system,
+                  cmd.target_component, cmd.command);
+
+        // Get the target component
+        // CameraComponent *tgt_comp = getCameraComponent(cmd.target_component);
```
Comment (lbegani) [inline src/mavlink_server.cpp]: Removed.

File: src/mavlink_server.cpp [hunk 8]
```diff
@@ -213,32 +285,183 @@ void MavlinkServer::_handle_camera_set_video_stream_settings(const struct sockad
...
+    bool ret = false;
+    mavlink_message_t msg2;
+    mavlink_param_ext_set_t param_set;
+    mavlink_param_ext_ack_t param_ext_ack;
+    mavlink_msg_param_ext_set_decode(msg, &param_set);
+    CameraComponent *tgtComp = getCameraComponent(param_set.target_component);
+    if (tgtComp) {
+        // Set parameter
+        ret = tgtComp->setParam(param_set.param_id, param_set.param_value, param_set.param_type);
+        strncpy(param_ext_ack.param_id, param_set.param_id, 16);
```
Comment (lbegani) [inline src/mavlink_server.cpp]: Fixed.

File: src/mavlink_server.cpp [hunk 9]
```diff
@@ -106,38 +107,107 @@ void MavlinkServer::_send_ack(const struct sockaddr_in &addr, int cmd, bool succ
...
+                                                       mavlink_command_long_t &cmd)
 {
+    log_debug("%s", __func__);
+
     mavlink_message_t msg;
     bool success = false;
 
-    if (action != 1) {
-        _send_ack(addr, command, true);
+    if (cmd.param1 != 1) {
```
Comment (lbegani) [inline src/mavlink_server.cpp]: Added comment for readability

File: src/CameraServer.cpp [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+#define DEVICE_PATH "/dev/"
+#define VIDEO_PREFIX "video"
+#define DEFAULT_SERVICE_PORT 8554
+
+CameraServer::CameraServer(ConfFile &conf)
+    : mavlink_server(conf, streams, rtsp_server)
+    , rtsp_server(streams, DEFAULT_SERVICE_PORT)
+    , cameraCount(0)
+{
+    cameraCount = detectCamera(conf);
```
Comment (lbegani) [inline src/CameraServer.cpp]: Idea of having CameraServer & CameraComponents is to serve requests from GCS like image capture, video capture, video streaming as well as camera control.  For now it only supports setting of some parameters for V4L2 camera device but going forward, support for realsense camera would also be added. Features like image capture and video capture will be added. Also, I plan to bring streaming under CameraServer eventually to have a coherent solution . For now I have kept the code independent of the existing streaming related classes which has resulted in duplication of camera detection (reading of device directory and blacklisting based on conf file).

File: src/CameraServer.cpp [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+    }
+
+    while (val_len > 0 && isspace(val[val_len - 1]))
+        val_len--;
+
+    if (val_len)
+        value_set->insert(std::string(val, val_len));
+}
+
+static int parse_stl_set(const char *val, size_t val_len, void *storage, size_t storage_len)
```
Comment (lbegani) [inline src/CameraServer.cpp]: I will check if this function can be moved to ConfFile class and can be used by both classes

File: src/main.cpp [hunk 12]
```diff
@@ -239,11 +242,17 @@ int main(int argc, char *argv[])
     conf = new ConfFile();
     parse_conf_files(*conf, &opt);
 
+#ifdef ENABLE_MAVLINK
+    CameraServer camServer(*conf);
```
Comment (lbegani) [inline src/main.cpp]: As explained above, I plan to do other way round. Add streaming feature to CameraComponent.

File: src/CameraComponent.h [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+    uint8_t storage_count;
+    uint8_t status;
+    float total_capacity;
+    float used_capacity;
+    float available_capacity;
+    float read_speed;
+    float write_speed;
+};
+
+class CameraComponent {
```
Comment (lbegani) [inline src/CameraComponent.h]: Explained above the reasoning behind CameraServer and CameraComponent classes.

File: src/mavlink_server.cpp (hunk 1 above)
Comment (otaviobp) [inline src/mavlink_server.cpp]: Well, anyway I don't like the idea of having a limit on the number of available cameras in the protocol. Even if it is big enough. But that's something to discuss with the dronecode community. 
And if I install camera-streaming-daemon in my product and I connect more than 6 usb cameras in it, it won't work. It is good to document that the camera-streaming-daemon will support only 6 cameras and inform users on how to change this if needed.

File: src/CameraServer.cpp (hunk 10 above)
Comment (otaviobp) [inline src/CameraServer.cpp]: I just don't get why to limit this functionality to V4L2 cameras and why not using the camera discovery mechanism that is used to create Streams.
Won't CameraServer support Realsense RGB, IR and depth streams?

File: src/CameraServer.cpp (hunk 11 above)
Comment (lbegani) [inline src/CameraServer.cpp]: Fixed, moved it to ConfFile class

File: src/mavlink_server.cpp (hunk 1 above)
Comment (lbegani) [inline src/mavlink_server.cpp]: Good Point, I will document that.

File: src/CameraServer.cpp (hunk 10 above)
Comment (lbegani) [inline src/CameraServer.cpp]: The functionality will not be limited to V4L2 cameras. Support will be added for real sense cameras too. However to do that, I first need to understand the controls and parameters of RS camera that can be exposed to user. Also, understand the APIs to  be able to set the parameters. Finally, make the changes. As a first step, decided to make the feature work for V4L2 and then take up RS.

File: src/CameraServer.cpp [hunk 14]
```diff
@@ -0,0 +1,172 @@
...
+        return 0;
+    }
+
+    log_debug("%s URI : %s", key, opt.uri_addr);
+    strcpy(value, opt.uri_addr);
+    free(opt.uri_addr);
+    return 1;
+}
+
+int CameraServer::detect_v4l2devices(ConfFile &conf, std::vector<CameraComponent *> &camList)
```
Comment (lbegani) [inline src/CameraServer.cpp]: Have now moved all v4l2 interfaces to a common file

File: src/CameraParameters.cpp [hunk 15]
```diff
@@ -97,6 +97,43 @@ bool CameraParameters::setParameter(std::string key, std::string value)
...
+    std::string str2(str, sizeof(str));
+    return setParameter(param_id, str2);
+}
+bool CameraParameters::setParameter(std::string param_id, uint32_t param_value)
+{
+    char str[CAM_PARAM_VALUE_LEN];
+    cam_param_union u;
+    u.param_uint32 = param_value;
+    memcpy(&str[0], &u.param_uint32, sizeof(uint32_t));
+    std::string str2(str, sizeof(str));
```
Comment (otaviobp) [inline src/CameraParameters.cpp]: This usage of c strings and c++ strings is incorrect. And you are going to lose information in some cases. And it is really hard to understand what are the reasons to do this.

For example if param_value is 0x4800 you are going to copy all those bytes to str (0x00, 0x48, 0x00, 0x00 - little endian) and than copy those bytes to str2 data. In CameraComponent_V4L2::getParam function you are going to call c_str() in str2.
As str2 first byte is 0, the c_str will be an empty string and no bytes will be copied.

File: src/CameraComponent_V4L2.cpp [hunk 16]
```diff
@@ -123,32 +129,40 @@ void CameraComponent_V4L2::initDefaultValues()
...
     std::string value = camParam.getParameter(param_id);
-    strcpy(param_value, value.c_str());
-    return 1;
+    if (!value.empty())
+        return 1;
+
+    if (value_size < value.size())
+        strncpy(param_value, value.c_str(), value_size);
+    else
+        strcpy(param_value, value.c_str());
```
Comment (otaviobp) [inline src/CameraComponent_V4L2.cpp]: This is unsafe. You can cause a buffer overflow or write less bytes than you should. I would review all usages of strcpy.

File: src/CameraParameters.cpp (hunk 15 above)
Comment (lbegani) [reply src/CameraParameters.cpp]: Why? The idea is to convert the parameter value from a data type to byte array and then store it as string.
It works, also I have a sample test app to test set/get operations. I will push that as one of the samples. Nevertheless, I have made changes to cut corner cases and simplify. Kindly review.

File: src/CameraComponent_V4L2.cpp (hunk 16 above)
Comment (lbegani) [reply src/CameraComponent_V4L2.cpp]: Fixed.

File: src/CameraParameters.cpp (hunk 15 above)
Comment (otaviobp) [reply src/CameraParameters.cpp]: I haven't seen the new code you updated, but I still believe that when using the code I reviewed you are going to lose information for some values of param_value and get invalid access memory for other values. Have you tried to use param_value=0x4800 as I suggested? As soon as you have a sample app that make it easy to test your data conversions, let me know.

Comment (lbegani) [reply src/CameraParameters.cpp]: I tried 0x4800 and it works. I have pushed the sample app(basic), you can try it out.

**Comment by lbegani** [comment]
@otaviobp @lfelipe @lucasdemarchi @anselmolsm Can you kindly review

**Comment by anitha-suresh** [comment]
@lbegani have we run coverity.
